ext/curl: Add `CURLOPT_PREREQFUNCTION`

Ayesh opened the pull request:
Curl >= 7.80.0 supports declaring a function for `CURLOPT_PREREQFUNCTION` option that gets called after Curl establishes a connection (including the TLS handshake for HTTPS connections), but before the actual request is made.

The callable must return either `CURL_PREREQFUNC_OK` or `CURL_PREREQFUNC_ABORT` to allow or abort the request.

This adds support for it to PHP with required ifdef.

 - libc: https://curl.se/libcurl/c/CURLOPT_PREREQFUNCTION.html

mvorisek commented on UPGRADING:
```diff
@@ -143,6 +143,12 @@ PHP 8.4 UPGRADE NOTES
+  . Added CURLOPT_PREREQFUNCTION option that accepts a callable that gets
+    called after a connection is established (including SSL handshake), but
+    before the request is made. The function must return CURL_PREREQFUNC_OK
+    or CURL_PREREQFUNC_OK to allow or abort the request.
```
```suggestion
    or CURL_PREREQFUNC_ABORT to allow or abort the request.
```

Ayesh replied:
Thank you, fixed now.

devnexen commented:
Looking good overall, @mvorisek remark is valid.

devnexen commented on ext/curl/tests/curl_setopt_CURLOPT_PREREQFUNCTION.phpt:
```diff
@@ -0,0 +1,50 @@
+
+$host = curl_cli_server_start();
+$url = "{$host}/get.inc?test=";
+curl_setopt($ch, CURLOPT_URL, $url);
```
do you think "enforcing" ip address kind with `CURLOPT_IPRESOLVE` would help ?

Ayesh replied:
Thank you so much! There was indeed a test failure because the test server had Curl server running on IPv6 loopback address. I tried "forcing" with IPv4, but had the tests failing for some reason. I ended up adding `filter_var` calls to check that the values passed are IP addresses, with tests passing now.

devnexen replied:
alright

SakiTakamachi commented:
In my opinion, `#if LIBCURL_VERSION_NUM >= 0x075000` is a hexadecimal literal, so it might be better to write the version in the comments as elsewhere.

Ayesh commented:
Thank you @SakiTakamachi - I added comments to indicate it's for Curl 7.80.0 throughout everywhere I had `ifdef` calls.

Girgias commented on ext/curl/curl_private.h:
```diff
@@ -75,6 +75,9 @@ typedef struct {
 	php_curl_callback *progress;
 #if LIBCURL_VERSION_NUM >= 0x072000
 	php_curl_callback  *xferinfo;
```
Note for self/refactoring, the php_curl_callback could all be replace with FCCs and the new FCC API I introduced as it keeps track of closures now.

Girgias commented on ext/curl/interface.c:
```diff
@@ -956,6 +962,66 @@ static size_t curl_write_header(char *data, size_t size, size_t nmemb, void *ctx
+	fci.named_params = NULL;
+
+	ch->in_callback = 1;
+	error = zend_call_function(&fci, &t->fci_cache);
```
You could just use ``zend_call_known_fcc()`` considering we already have the FCC.

Girgias commented on ext/curl/interface.c:
```diff
@@ -956,6 +962,66 @@ static size_t curl_write_header(char *data, size_t size, size_t nmemb, void *ctx
+	ch->in_callback = 1;
+	error = zend_call_function(&fci, &t->fci_cache);
+	ch->in_callback = 0;
+	if (error == FAILURE) {
+		php_error_docref(NULL, E_WARNING, "Cannot call the CURLOPT_PREREQFUNCTION");
+	} else if (!Z_ISUNDEF(retval)) {
```
This should never happen, ``zend_call_function()`` only returns FAILURE if the executor is not running.

Girgias commented:
Can you wait for #13291 and rebase on top of this when it is merged? As the refactoring should simplify the code.

Ayesh commented:
Awesome, thank you! I can also learn how to do it from your PR too, so this is perfect.

Ayesh commented:
@Girgias now that #13291 is merged, I rebased and changed this to use your new FCC pattern. Could you review when you get some time please? Thank you.

Girgias commented on ext/curl/interface.c:
```diff
@@ -711,6 +716,53 @@ static size_t curl_xferinfo(void *clientp, curl_off_t dltotal, curl_off_t dlnow,
+			if (retval_long == CURL_PREREQFUNC_OK || retval_long == CURL_PREREQFUNC_ABORT) {
+				rval = retval_long;
+			} else {
+				zend_throw_error(NULL, "The CURLOPT_PREREQFUNCTION callback must return either CURL_PREREQFUNC_OK or CURL_PREREQFUNC_ABORT");
```
ValueError?

Ayesh replied:
Yeah I think you are right - changed to a `ValueError`. The error message text is the same as the `TypeError` check, but now we throw `ValueError` or `TypeError` depending on the case.

Girgias commented on ext/curl/interface.c:
```diff
@@ -711,6 +716,53 @@ static size_t curl_xferinfo(void *clientp, curl_off_t dltotal, curl_off_t dlnow,
+				zend_throw_error(NULL, "The CURLOPT_PREREQFUNCTION callback must return either CURL_PREREQFUNC_OK or CURL_PREREQFUNC_ABORT");
+			}
+		} else {
+			zend_throw_error(NULL, "The CURLOPT_PREREQFUNCTION callback must return either CURL_PREREQFUNC_OK or CURL_PREREQFUNC_ABORT");
```
This is a TypeError as the type returned by the callable is not an int.

Ayesh replied:
Thank you, changed to a `TypeError` for non-int return values.

Girgias commented on ext/curl/tests/curl_setopt_CURLOPT_PREREQFUNCTION.phpt:
```diff
@@ -0,0 +1,61 @@
+$ch = curl_init();
+curl_setopt($ch, CURLOPT_URL, $host);
+curl_setopt($ch, CURLOPT_RETURNTRANSFER, 1);
+//curl_setopt($ch, CURLOPT_IPRESOLVE, CURL_IPRESOLVE_V4);
```
Why is this commented?

Ayesh replied:
I think I was trying to fix the cli server connections were using IPv6 loopback address, so my previous test that expected an IPv4 address failed. I replaced with a `filter_var` / `FILTER_VALIDATE_IP` call, so it wasn't needed. I cleaned it up.

Girgias reviewed (commented):
Minor nits, can you also add a trampoline callback test case?

Girgias commented:
@Ayesh ping

Girgias reviewed (approved):
LGTM if CI is green

Ayesh commented:
Thank you @Girgias - I made a bunch of force-pushes with a few more added tests, but I tested them locally. fingers crossed CI marks them green too.

Ayesh commented:
`curl_setopt($ch, CURLOPT_PREREQFUNCTION, null)` seems to segfault, I will sort out and fix soon :(

Girgias commented on ext/curl/tests/curl_setopt_CURLOPT_PREREQFUNCTION.phpt:
```diff
@@ -0,0 +1,152 @@
+	var_dump(func_num_args());
+	$args = func_get_args();
+	var_dump(get_class($args[0]));
+	var_dump(filter_var($args[1], FILTER_VALIDATE_IP) !== false);
+	var_dump(filter_var($args[2], FILTER_VALIDATE_IP) !== false);
```
Would it be possible to have the test not rely on the filter extension? Found this out while trying it locally without ext/filter compiled.

Otherwise it needs to be added to the EXTENSIONS section as I did.

Ayesh commented:
CI green finally, and `curl_setopt($ch, CURLOPT_PREREQFUNCTION, null); curl_exec($ch);` works now :)
